app/services/CheckoutService.scala

```
@@ -42,101 +42,101 @@ class CheckoutService(identityService: IdentityService,
-      subscribe <- EitherT(createSubscribeRequest(personalData, requestData, plan, userData, memberId, subscriptionData, payment))
+      memberId <- EitherT(createOrUpdateUserInSalesforce(personalData, idMinimalUser, subscriptionData))
+      purchaserIds = PurchaserIdentifiers(memberId, idMinimalUser)
+      payment <- EitherT(createPaymentType(personalData, requestData, purchaserIds, subscriptionData))
+      subscribe <- EitherT(createSubscribeRequest(personalData, requestData, plan, purchaserIds, subscriptionData, payment))
       withPromo = promoService.applyPromotion(subscribe, subscriptionData.suppliedPromoCode, Some(personalData.country))
-      result <- EitherT(createSubscription(withPromo, userData, subscriptionData))
-      _ <- EitherT(sendETDataExtensionRow(result, subscriptionData, gracePeriod(subscribe, withPromo), userData))
-      _ <- EitherT(identityUpdate)
-    } yield {
-      CheckoutSuccess(memberId, userData, result, subscribe.promoCode)
-    }).run
+      result <- EitherT(createSubscription(withPromo, purchaserIds, subscriptionData))
+      _ <- EitherT(sendETDataExtensionRow(result, subscriptionData, gracePeriod(subscribe, withPromo), purchaserIds))
+      identitySuccess <- storeIdentityDetails(personalData, authenticatedUserOpt, memberId)
+    } yield CheckoutSuccess(memberId, identitySuccess.userData, result, subscribe.promoCode)).run
+  }
+
+  private def storeIdentityDetails(personalData: PersonalData, authenticatedUserOpt: Option[AuthenticatedIdUser], memberId: ContactId): EitherT[Future, NonEmptyList[SubsError], IdentitySuccess] = {
+    authenticatedUserOpt match {
+      case Some(authenticatedIdUser) => {
+        EitherT(updateAuthenticatedUserDetails(authenticatedIdUser, memberId, personalData))
+      }
+      case None => {
+        logger.info(s"User does not have an Identity account. Creating a guest account")
+        for (identitySuccess <- EitherT(identityService.registerGuest(personalData))) yield {
```

could you replace these two nested for statements with one for comprehension?
I assume we want to swallow & log both identity reg errors and salesforce update errors